export/mlab_export.py

```diff
@@ -0,0 +1,591 @@
+    values: iterable of float, values corresponding to each timestamp.
+  Returns:
+    list of dict, each dict has keys timestamp and value.
+  """
+  samples = []
+  assert(len(timestamps) == len(values))
+  for i in xrange(0, len(timestamps)):
+    if values[i] is not None:
+      samples.append({'timestamp': timestamps[i], 'value': values[i]})
+  return samples
+
+
+def write_json_record(fd_output, record, pretty_json):
+  """Writes json record to fd_output.
+
+  Args:
+    fd_output: file object open for writing, the record is written to this fd.
+    record: dict, the record of data to serialize as json.
+    pretty_json: bool, whether to write the json with extra spacing.
+  """
+  json.dump(record, fd_output, indent=pretty_json)
+  fd_output.write('\n')  # separate each record with newline.
+
+
+def get_rrd_files(rrddir_prefix):
+  """Returns the absolute path of all rrd files found under rrddir_prefix.
+
+  Args:
+    rrddir_prefix: str, base directory where rrd files are stored.
+  Returns:
+    list of str, where each element is the absolute path to a single rrd file.
+  """
+  rrdfiles = []
+  for root, _, filenames in os.walk(rrddir_prefix):
+    for filename in filenames:
+      file_ext = os.path.join(root, filename)
+      if file_ext.endswith('.rrd'):
+        rrdfiles.append(os.path.abspath(file_ext))
+  return rrdfiles
+
+
+def rrd_list(options):
+  """Processes all options.show_* flags without performing an export."""
+  for filename in get_rrd_files(options.rrddir_prefix):
+    (_, value_names, _) = rrdtool.fetch(
+        filename, 'AVERAGE', '--start', str(options.ts_start), '--end',
+        str(options.ts_end))
+    for value_name in value_names:
+      get_canonical_names(filename, value_name, options)
+
+
+def rrd_export(options):
+  """Exports all RRD data."""
+  open_func = open
+  if options.compress:
+    open_func = gzip.open
+    options.output += '.gz'
+  fd_output = open_func(options.output, 'w')
+  for filename in get_rrd_files(options.rrddir_prefix):
+    (time_range, value_names, data) = rrdtool.fetch(
+        filename, 'AVERAGE', '--start', str(options.ts_start), '--end',
+        str(options.ts_end))
+    # W0142 is the use of "* magic". These are legitimate use-cases.
+    # 1) time_range is a 3-tuple (start, end, step): i.e. arguments to range.
+    # 2) data is a list of tuples, which are transposed by zip.
+    #    i.e. [(a,), (b,), ...] -> [(a,b,...)]
+    # pylint: disable=W0142
+    timestamps = range(*time_range)
+    values = zip(*data)
+    # pylint: enable=W0142
+
+    for i, value_name in enumerate(value_names):
+      hostname, experiment, metric = get_canonical_names(
+          filename, value_name, options)
+      if metric is None or experiment in options.ignored_experiments:
+        continue
+      record = get_json_record(
+          hostname, experiment, metric, timestamps, values[i])
+      write_json_record(fd_output, record, options.pretty_json)
+  fd_output.close()
```

Comment:
I wondered that originally, and discovered that gzip.open was not context-aware object. But, I looked for other options based on your question and discovered contextlib, which actually makes this possible. Yay.

How does it look? 11ec1837d6a0b68437d118fda689c2d75adef1f9